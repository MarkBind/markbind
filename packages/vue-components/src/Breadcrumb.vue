<template>
  <div>
    <nav aria-label="breadcrumb" class="breadcrumb-divider">
      <ol class="breadcrumb">
        <li class="notlink breadcrumb-marker">
          ››&nbsp;&nbsp;
        </li>
        <li
          v-for="(item, index) in items"
          :key="index"
          :class="['breadcrumb-item',
                   {'notlink': item.link === null},
          ]"
          :aria-current="{'page': isLast(index, items.length)}"
        >
          <a v-if="item.link === null">
            {{ item.title }}
          </a>
          <a v-else :href="item.link">
            {{ item.title }}
          </a>
        </li>
      </ol>
    </nav>
  </div>
</template>

<script>
export default {
  data() {
    return {
      items: [],
    };
  },
  methods: {
    isLast(index, length) {
      return index === length - 1;
    },
  },
  mounted() {
    // Get root of site-nav
    const siteNav = Array.from(document.querySelectorAll('ul')).find(el =>
      el.classList.contains('site-nav-list-root'));

    if (!siteNav) return;

    // Find current link in site-nav
    const currLink = siteNav.querySelector('.current');

    if (!currLink) return;

    this.items.unshift({
      'title': currLink.textContent,
      'link': null,
    });

    let currentEl = currLink.parentElement;

    while (currentEl !== siteNav) {
      if (currentEl.tagName.toLowerCase() === 'ul') {
        const divElement = currentEl.parentElement.querySelector('div');
        const aElement = divElement.querySelector('a[href]');
        // if does not contain link
        if (aElement === null) {
          this.items.unshift({
            'title': divElement.textContent,
            'link': null,
          });
        } else {
          this.items.unshift({
            'title': aElement.textContent,
            'link': aElement.getAttribute('href'),
          });
        }
      }
      currentEl = currentEl.parentElement;
    }
  },
};
</script>

<style scoped>
    .notlink {
        color: #6d757d;
    }

    .breadcrumb-divider {
        --bs-breadcrumb-divider: '›';
    }

    /* Change font size to center arrows */
    .breadcrumb-marker {
        font-size: 15px;
    }

    .breadcrumb-item::before {
        font-size: 15px;
    }
</style>
