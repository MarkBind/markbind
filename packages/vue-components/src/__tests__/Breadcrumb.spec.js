import { nextTick } from 'vue';
import { mount } from '@vue/test-utils';
import Breadcrumb from '../Breadcrumb.vue';

const SITENAV_START = `
<nav id="site-nav">
<div class="nav-component slim-scroll">
<div class="site-nav-root">
<ul class="site-nav-list site-nav-list-root">
`;

const SITENAV_END = `
</ul>
</div>
</div>
</nav>
`;

const SITENAV_LINK = `
<li>
<div class="site-nav-default-list-item">
<a href="#link" class="current">
Link
</a>
</div>
</li>
`;

const oneLinkSiteNav = {
  template: `
  ${SITENAV_START}
  ${SITENAV_LINK}
  ${SITENAV_END}
  `,
};

const dropdownNonLinkSiteNav = {
  template: `
  ${SITENAV_START}
  <li>
  <div class="site-nav-default-list-item">
  Non-Link Dropdown
  </div>
  <ul>
  ${SITENAV_LINK}
  </ul>
  </li>
  ${SITENAV_END}
  `,
};

const dropdownLinkSiteNav = {
  template: `
  ${SITENAV_START}
  <li>
  <div class="site-nav-default-list-item">
  <a href="#link-dropdown">
  Link Dropdown
  </a>
  </div>
  <ul>
  ${SITENAV_LINK}
  </ul>
  </li>
  ${SITENAV_END}
  `,
};

describe('Breadcrumb', () => {
  test('empty breadcrumb', async () => {
    const wrapper = mount(Breadcrumb, {
      data() {
        return {
          items: [],
        };
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  test('one non-link', async () => {
    const wrapper = mount(Breadcrumb, {
      data() {
        return {
          items: [
            {
              'title': 'non-link one',
              'link': null,
            },
          ],
        };
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  test('one link', async () => {
    const wrapper = mount(Breadcrumb, {
      data() {
        return {
          items: [
            {
              'title': 'link one',
              'link': '#link',
            },
          ],
        };
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  test('two non-link', async () => {
    const wrapper = mount(Breadcrumb, {
      data() {
        return {
          items: [
            {
              'title': 'link one',
              'link': null,
            },
            {
              'title': 'link two',
              'link': null,
            },
          ],
        };
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  test('two link', async () => {
    const wrapper = mount(Breadcrumb, {
      data() {
        return {
          items: [
            {
              'title': 'link one',
              'link': '#linkone',
            },
            {
              'title': 'link two',
              'link': '#linktwo',
            },
          ],
        };
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  test('no sitenav', async () => {
    const wrapper = mount(Breadcrumb, {
      attachTo: document.body,
    });

    expect(wrapper.element).toMatchSnapshot();
    wrapper.unmount();
  });

  test('one link from sitenav', async () => {
    const sitenav = mount(oneLinkSiteNav, {
      attachTo: document.body,
    });
    const wrapper = mount(Breadcrumb, {
      attachTo: document.body,
    });
    await nextTick();

    expect(wrapper.element).toMatchSnapshot();
    sitenav.unmount();
    wrapper.unmount();
  });

  test('dropdown non-link from sitenav', async () => {
    const sitenav = mount(dropdownNonLinkSiteNav, {
      attachTo: document.body,
    });
    const wrapper = mount(Breadcrumb, {
      attachTo: document.body,
    });
    await nextTick();

    expect(wrapper.element).toMatchSnapshot();
    sitenav.unmount();
    wrapper.unmount();
  });

  test('dropdown link from sitenav', async () => {
    const sitenav = mount(dropdownLinkSiteNav, {
      attachTo: document.body,
    });
    const wrapper = mount(Breadcrumb, {
      attachTo: document.body,
    });
    await nextTick();

    expect(wrapper.element).toMatchSnapshot();
    sitenav.unmount();
    wrapper.unmount();
  });
});
