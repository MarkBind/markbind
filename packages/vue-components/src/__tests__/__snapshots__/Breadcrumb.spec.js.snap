// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumb dropdown link from sitenav 1`] = `
<div>
  <nav
    aria-label="breadcrumb"
    class="breadcrumb-divider"
  >
    <ol
      class="breadcrumb"
    >
      <li
        class="notlink breadcrumb-marker"
      >
         ››   
      </li>
      
      <li
        aria-current="[object Object]"
        class="breadcrumb-item"
      >
        <a
          href="#link-dropdown"
        >
           Link Dropdown 
        </a>
      </li>
      <li
        aria-current="[object Object]"
        class="breadcrumb-item notlink"
      >
        <a>
           Link 
        </a>
      </li>
      
    </ol>
  </nav>
</div>
`;

exports[`Breadcrumb dropdown non-link from sitenav 1`] = `
<div>
  <nav
    aria-label="breadcrumb"
    class="breadcrumb-divider"
  >
    <ol
      class="breadcrumb"
    >
      <li
        class="notlink breadcrumb-marker"
      >
         ››   
      </li>
      
      <li
        aria-current="[object Object]"
        class="breadcrumb-item notlink"
      >
        <a>
           Non-Link Dropdown 
        </a>
      </li>
      <li
        aria-current="[object Object]"
        class="breadcrumb-item notlink"
      >
        <a>
           Link 
        </a>
      </li>
      
    </ol>
  </nav>
</div>
`;

exports[`Breadcrumb empty breadcrumb 1`] = `
<div>
  <nav
    aria-label="breadcrumb"
    class="breadcrumb-divider"
  >
    <ol
      class="breadcrumb"
    >
      <li
        class="notlink breadcrumb-marker"
      >
         ››   
      </li>
      
      
    </ol>
  </nav>
</div>
`;

exports[`Breadcrumb no sitenav 1`] = `
<div>
  <nav
    aria-label="breadcrumb"
    class="breadcrumb-divider"
  >
    <ol
      class="breadcrumb"
    >
      <li
        class="notlink breadcrumb-marker"
      >
         ››   
      </li>
      
      
    </ol>
  </nav>
</div>
`;

exports[`Breadcrumb one link 1`] = `
<div>
  <nav
    aria-label="breadcrumb"
    class="breadcrumb-divider"
  >
    <ol
      class="breadcrumb"
    >
      <li
        class="notlink breadcrumb-marker"
      >
         ››   
      </li>
      
      <li
        aria-current="[object Object]"
        class="breadcrumb-item"
      >
        <a
          href="#link"
        >
          link one
        </a>
      </li>
      
    </ol>
  </nav>
</div>
`;

exports[`Breadcrumb one link from sitenav 1`] = `
<div>
  <nav
    aria-label="breadcrumb"
    class="breadcrumb-divider"
  >
    <ol
      class="breadcrumb"
    >
      <li
        class="notlink breadcrumb-marker"
      >
         ››   
      </li>
      
      <li
        aria-current="[object Object]"
        class="breadcrumb-item notlink"
      >
        <a>
           Link 
        </a>
      </li>
      
    </ol>
  </nav>
</div>
`;

exports[`Breadcrumb one non-link 1`] = `
<div>
  <nav
    aria-label="breadcrumb"
    class="breadcrumb-divider"
  >
    <ol
      class="breadcrumb"
    >
      <li
        class="notlink breadcrumb-marker"
      >
         ››   
      </li>
      
      <li
        aria-current="[object Object]"
        class="breadcrumb-item notlink"
      >
        <a>
          non-link one
        </a>
      </li>
      
    </ol>
  </nav>
</div>
`;

exports[`Breadcrumb two link 1`] = `
<div>
  <nav
    aria-label="breadcrumb"
    class="breadcrumb-divider"
  >
    <ol
      class="breadcrumb"
    >
      <li
        class="notlink breadcrumb-marker"
      >
         ››   
      </li>
      
      <li
        aria-current="[object Object]"
        class="breadcrumb-item"
      >
        <a
          href="#linkone"
        >
          link one
        </a>
      </li>
      <li
        aria-current="[object Object]"
        class="breadcrumb-item"
      >
        <a
          href="#linktwo"
        >
          link two
        </a>
      </li>
      
    </ol>
  </nav>
</div>
`;

exports[`Breadcrumb two non-link 1`] = `
<div>
  <nav
    aria-label="breadcrumb"
    class="breadcrumb-divider"
  >
    <ol
      class="breadcrumb"
    >
      <li
        class="notlink breadcrumb-marker"
      >
         ››   
      </li>
      
      <li
        aria-current="[object Object]"
        class="breadcrumb-item notlink"
      >
        <a>
          link one
        </a>
      </li>
      <li
        aria-current="[object Object]"
        class="breadcrumb-item notlink"
      >
        <a>
          link two
        </a>
      </li>
      
    </ol>
  </nav>
</div>
`;
